src/main/kotlin/org/rust/ide/inspections/RsVariableMutableInspection.kt

```diff
@@ -0,0 +1,59 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.LocalSearchScope
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.annotator.fixes.RemoveMutableFix
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.isMut
+import org.rust.lang.core.psi.ext.parentOfType
+import org.rust.lang.core.psi.ext.selfParameter
+import org.rust.lang.core.resolve.ref.RsMethodCallReferenceImpl
+
+class RsVariableMutableInspection : RsLocalInspectionTool() {
+    override fun getDisplayName() = "No mutable required"
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) =
+        object : RsVisitor() {
+            override fun visitPatBinding(o: RsPatBinding) {
+                if (!o.isMut) return
+                val block = o.parentOfType<RsBlock>() ?: o.parentOfType<RsFunction>() ?: return
+                val occurrences = ReferencesSearch.search(o, LocalSearchScope(block))
+                    .map { it.element.parent }
+                occurrences
+                    .asSequence()
+                    .filterNot { checkOccurrenceNotNeedMutable(it) }
+                    .forEach { return }
+                holder.registerProblem(o, "No mutable required for ${o.identifier.text}", *arrayOf(RemoveMutableFix(o)))
+            }
+        }
+
+    fun checkOccurrenceNotNeedMutable(occurrence: PsiElement): Boolean {
+        val parent = occurrence.parent
+        when (parent) {
+            is RsUnaryExpr -> if (handleUnaryExpr(parent)) {
+                return false
+            }
+            is RsBinaryExpr -> if (parent.left == occurrence) {
+                return false
+            }
+            is RsMethodCallExpr -> {
+                val ref = parent.reference.resolve() as? RsFunction ?: return false
+                val self = ref.selfParameter ?: return false
+                if (self.isMut) {
+                    return false
+                }
+            }
+            is RsTupleExpr,
```

Comment:
I don't remember if this case was in your previous version, but actually, just checking for `RsTupleExpr` at a particular level doesn't change much. You need to traverse ancestors of your occurrence up while they are tuples and check mutability at every level. As far as I understand, the following simple code will emit false positive:

```rust
fn foo(a: &mut (u8)) {}
fn main() {
    let mut a = 10;
    foo(&mut (a));
}
```